### ISSUE TITLE
Notebook Questions

### ISSUE THREAD

**Issue by TonyCostanzo** [IE-482-582/fall2021#8]
Open thread for all of us to post questions we've recorded in our student_workbook.md files.

**Comment by TonyCostanzo** [comment]
### Chapter 1 -- Introduction

- #### Unanswered Questions:

> 1) I guess I don't fully understand what ROS is phsyically. Is it a database stored somewhere with heaps of code similar to a 'Stack Overflow, is it some massive synthesizer that utilizes a series of different compilers to allow code from a multitude of different languages? Simply put, what does "a framework" mean physically in regard to ROS?

### Chapter 2 -- Preliminaries

- #### Unanswered Questions:

> 1) Is the running 'roscore' unique to the machine it was started on? 

> 2) How does the 'listener' node know to listen to 'talker' or is 'listener' just a generic listening node that would receive input from anything chatting?

### Chapter 3 -- Topics

- #### Unanswered Questions:

> 1) (Pg. 33) What does it mean we have to add a dependency to the package.xml file? Is this done concurrently with writing this example node? If so where does this dependdency tag go?

> 2) What does: "Usage: rosrun [--prefix cmd] [--debug] PACKAGE EXECUTABLE [ARGS]
  rosrun will locate PACKAGE and try to find
  an executable named EXECUTABLE in the PACKAGE tree.
  If it finds it, it will run it with ARGS." mean in plain English?

**Comment by Danealor** [comment]
Ah yes, I will post some of my own questions here as well. In the meantime, I think I can answer a few of @TonyCostanzo's questions with the knowledge I have so far.

## Chapter 1
1. From what I understand, ROS as a robotics framework acts in a similar manner as Unity or Unreal Engine do as game development frameworks. It is a set of code, present on your machine I believe in `/opt/ros/noetic` as it were installed. This code and programs are responsible for mediating the registrations of the nodes within the network, tying different modules together and generally providing an incredibly extensive boilerplate base that allows us to build up our robot code almost like lego pieces. I believe you are correct about the compilation part, I think that's `catkin`'s job mostly.

## Chapter 2
2. The 'listener' node registers for a message type (topic) called "chatter" with a data type of String. The 'talker' node publishes a topic called "chatter". Like fate-bound lovers they find each other (through `roscore` giving each of them the others' address) and start talking to each other.

In `listener.py`:
```
rospy.Subscriber("chatter", String, callback)
```

**Comment by jhulburt** [comment]
## Textbook Chapters

### Chapter 1 -- Introduction

Just read this chapter...**DO NOT INSTALL ANYTHING**. Your virtual machine already has the software loaded and configured for you.  


- Unanswered Questions:
	- none pretty general overview
	


---

### Chapter 2 -- Preliminaries

Just read this chapter...**DO NOT INSTALL ANYTHING**. Your virtual machine already has the software loaded and configured for you.  

- Unanswered Questions:
	- What does a bash file do?
	- What are some examples of specific "parameters"





---

### Chapter 3 -- Topics

- Unanswered Questions:
	- what is the chmod function and what do its different flags denote
	- more background on Cmake as well?
	- will we define or own message type often?

### Chapter 4 -- Services	

- Unanswered Questions:
	- methods to tel a Node to shut down, if not with roscore than with what?
	- why is a service different from anyother node? when would you use it?
	


---

### Chapter 5 -- Actions

- Unanswered Questions:
	
- Answered Questions and/or General notes
	- Actions: used in extended time, goal oriented behaviors where constant feedback is neccesary. like movement
				a higher level protocol that specifies how topics should be used in combination
		- Arguments: goal, result feedback
		- Implement: Initialize and start action server node
					 define name, action type, goal, then spin it

**Comment by TonyCostanzo** [comment]
### Chapter 4 -- Services

- #### Questions:

> Q 4.1) On page 55, the callback function evidently takes type `WordCountRequest`, but the script     only has `request` in the parameters of the `count_words` function call. How do these two types relate?
